daphne/src/roles/helper.rs

```diff
@@ -241,39 +243,38 @@ pub trait DapHelper<S>: DapAggregator<S> {
             },
         )?;
         let part_batch_sel = state.part_batch_sel.clone();
-        let transition = task_config.vdaf.handle_agg_job_cont_req(
+
+        let num_reports = state.seq.len();
+        let transitions_and_shares = task_config.vdaf.handle_agg_job_cont_req(
             task_id,
             &agg_job_id,
             state,
             &agg_job_cont_req,
             &metrics,
         )?;
 
-        let (agg_job_resp, out_shares_count) = match transition {
-            DapHelperTransition::Continue(..) => {
-                return Err(fatal_error!(err = "unexpected transition (continued)").into());
-            }
-            DapHelperTransition::Finish(out_shares, mut agg_job_resp) => {
-                let out_shares_count = u64::try_from(out_shares.len()).unwrap();
+        let mut transitions = Vec::with_capacity(num_reports);
+        let mut out_shares_count = 0;
+
+        for transition_and_share in transitions_and_shares {
+            let (mut transition, share) = transition_and_share?;
+            if let Some(share) = share {
+                out_shares_count += 1;
                 let replayed = self
-                    .put_out_shares(task_id, task_config, &part_batch_sel, out_shares)
+                    .put_out_shares(task_id, task_config, &part_batch_sel, vec![share])
```

Comment:
Hmm, this doesn't look like it'll be very efficient for `daphne_worker`. What about `put_out_shares()` an iterator that we filter here (`replayed.contains(&transiition.report_id`)